### PR TITLE
[SPARK-20894][SS]Resolve the checkpoint location in driver and use the resolved path in state store

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

When the user runs a Structured Streaming query in a cluster, if the driver uses the local file system, StateStore running in executors will throw a file-not-found exception. However, the current error is not obvious.

This PR makes StreamExecution resolve the path in driver and uses the full path including the scheme part (such as `hdfs:/`, `file:/`) in StateStore.

Then if the above error happens, StateStore will throw an error with this full path which starts with `file:/`, and it makes this error obvious: the checkpoint location is on the local file system.

One potential minor issue is that the user cannot use different default file system settings in driver and executors (e.g., use a public HDFS address in driver and a private HDFS address in executors) after this change. However, since the batch query also has this issue (See https://github.com/apache/spark/blob/4bb6a53ebd06de3de97139a2dbc7c85fc3aa3e66/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L402), it doesn't make things worse.

## How was this patch tested?

The new added test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -595,6 +601,7 @@ class StreamExecution(
 
   /**
    * Processes any data available between `availableOffsets` and `committedOffsets`.
+   *
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: extra line

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala [hunk 2]
```diff
@@ -614,6 +614,24 @@ class StreamSuite extends StreamTest {
...
   }
+
+  test("should resolve the checkpoint path") {
+    withTempDir { dir =>
+      val checkpointLocation = dir.getCanonicalPath
+      assert(!checkpointLocation.startsWith("file:/"))
+      val query = MemoryStream[Int].toDF
+        .writeStream
+        .option("checkpointLocation", checkpointLocation)
+        .format("console").start()
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala]: .start() on the next line.

**Comment by tdas** [comment]
roughly LGTM, as long as you resolve conflicts and tests pass.

**Comment by zsxwing** [comment]
Thanks! Merging to master and 2.2.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala [hunk 3]
```diff
@@ -617,6 +617,25 @@ class StreamSuite extends StreamTest {
     query.stop()
   }
 
+  test("should resolve the checkpoint path") {
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala]: Just test conflicts.

**Pull request closed without merging**